### PR TITLE
Support EIP4844 Blobs indexing and API

### LINKED ISSUE #8602: Support EIP-4844
https://eips.ethereum.org/EIPS/eip-4844

### PR DESCRIPTION
**Pull request by k1rill-fedoseev**
Closes #8602

## Changelog

* Add tables for storing eip4844 transaction/block fields and data blobs themselves
  * New transaction fields are stored in a separate extension table and accessed using preloads and joins to avoid storing null entries for 99% of transactions
* Add `/api/v2/transactions/:transaction_hash_param/blobs`, `/api/v2/blobs/:blob_hash_param`
* Support `/api/v2/transactions?type=[blob_transaction]` transaction filter for blobs page (there is no `/api/v2/blobs` list api for now)
* Add new `blob_versioned_hashes`, `blob_gas_price`, `blob_gas_used`, `max_fee_per_blob_gas` in `/api/v2/transactions`, `/api/v2/transactions/:transaction_hash` and `/api/v2/blocks/:block_hash_or_number/transactions` responses
* Add new `blob_gas_price`, `blob_gas_used`, `excess_blob_gas`, `burnt_blob_fees` in`/api/v2/blocks/:block_hash_or_number` response
* Add new `blob_gas_used`, `excess_blob_gas` in`/api/v2/blocks` response
* Add data blobs indexer for fetching blobs from the beacon chain node API

All changes mentioned above apply only when using `ethereum` chain type on eip4844-supporting instances (ethereum, gnosis, lukso and their respective testnets).

New envs: https://github.com/blockscout/docs/pull/231

## Checklist for your Pull Request (PR)

  - [x] I added an entry to `CHANGELOG.md` with this PR
  - [x] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
  - [x] I checked whether I should update the docs and did so by submitting a PR to https://github.com/blockscout/docs
  - [x] If I added/changed/removed ENV var, I submitted a PR to https://github.com/blockscout/docs to update the list of env vars at https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables.md and I updated the version to `master` in the Version column. Changes will be reflected in this table: https://docs.blockscout.com/for-developers/information-and-settings/env-variables. 
  - [ ] If I add new indices into DB, I checked, that they are not redundant with PGHero or other tools

### CODE REVIEW

File: apps/block_scout_web/lib/block_scout_web/views/api/v2/transaction_view.ex [hunk 1]
```diff
@@ -449,9 +449,25 @@ defmodule BlockScoutWeb.API.V2.TransactionView do
 
         Map.put(extended_result, "zkevm_status", zkevm_status(extended_result))
 
-      "suave" ->
+      {true, "suave"} ->
         suave_fields(transaction, result, single_tx?, conn, watchlist_names)
```
Comment (k1rill-fedoseev) [inline apps/block_scout_web/lib/block_scout_web/views/api/v2/transaction_view.ex:453]: There are some new dialyzer warnings here regarding `single_tx?` being always true, although functionally code should not have changed

File: apps/block_scout_web/lib/block_scout_web/views/api/v2/blob_view.ex [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  def prepare_blob(blob) do
+    %{
+      "hash" => blob.hash,
+      "blob_data" => encode_binary(blob.blob_data),
+      "kzg_commitment" => encode_binary(blob.kzg_commitment),
+      "kzg_proof" => encode_binary(blob.kzg_proof)
+    }
+  end
+
+  defp encode_binary(binary) do
```
Comment (vbaranov) [inline apps/block_scout_web/lib/block_scout_web/views/api/v2/blob_view.ex:29]: There is function with the same logic in https://github.com/blockscout/blockscout/blob/master/apps/indexer/lib/indexer/transform/token_transfers.ex#L331-L333. Could we re-use it?

File: apps/explorer/lib/explorer/chain/beacon/reader.ex [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+      select: 3
+    ]
+
+  import Explorer.Chain, only: [select_repo: 1]
+
+  alias Explorer.Chain.Beacon.{Blob, BlobTransaction}
+  alias Explorer.{Chain, PagingOptions, Repo}
+  alias Explorer.Chain.{Hash, Transaction}
+
+  def blob(hash, options) when is_list(options) do
```
Comment (vbaranov) [inline apps/explorer/lib/explorer/chain/beacon/reader.ex:23]: Please add spec & doc for public function

File: apps/explorer/lib/explorer/chain/beacon/reader.ex [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    Blob
+    |> where(hash: ^hash)
+    |> select_repo(options).one()
+    |> case do
+      nil -> {:error, :not_found}
+      batch -> {:ok, batch}
+    end
+  end
+
+  def blob_hash_to_transactions(hash, options) when is_list(options) do
```
Comment (vbaranov) [inline apps/explorer/lib/explorer/chain/beacon/reader.ex:33]: Please add spec & doc for public function

File: apps/explorer/lib/explorer/chain/beacon/reader.ex [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    |> order_by([bt, transaction], desc: transaction.block_consensus, desc: transaction.block_number)
+    |> limit(10)
+    |> select([bt, transaction], %{
+      block_consensus: transaction.block_consensus,
+      transaction_hash: transaction.hash
+    })
+    |> select_repo(options).all()
+  end
+
+  def stream_missed_blob_transactions_timestamps(min_block, max_block, initial, reducer, options \\ [])
```
Comment (vbaranov) [inline apps/explorer/lib/explorer/chain/beacon/reader.ex:46]: Please add spec & doc for public function

File: apps/explorer/lib/explorer/chain/beacon/reader.ex [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            }
+          )
+        ),
+        inner_join: transaction in Transaction,
+        on: transaction_blob.transaction_hash == transaction.hash,
+        where: transaction.type == 3,
+        left_join: blob in Blob,
+        on: blob.hash == transaction_blob.blob_hash,
+        where: is_nil(blob.hash),
+        distinct: transaction.block_timestamp,
```
Comment (vbaranov) [inline apps/explorer/lib/explorer/chain/beacon/reader.ex:65]: When denormalization is not finished, we need to join with `blocks` table in order to get `block.timestamp`. Usually, we check it with `if DenormalizationHelper.denormalization_finished?() ...`.

File: apps/explorer/lib/explorer/chain/beacon/reader.ex [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+  end
+
+  def blob_hash_to_transactions(hash, options) when is_list(options) do
+    BlobTransaction
+    |> where(type(^hash, Hash.Full) == fragment("any(blob_versioned_hashes)"))
+    |> join(:inner, [bt], transaction in Transaction, on: bt.hash == transaction.hash)
+    |> order_by([bt, transaction], desc: transaction.block_consensus, desc: transaction.block_number)
+    |> limit(10)
+    |> select([bt, transaction], %{
+      block_consensus: transaction.block_consensus,
```
Comment (vbaranov) [inline apps/explorer/lib/explorer/chain/beacon/reader.ex:40]: When denormalization is not finished, we need to join with `blocks` table in order to get block.consensus.

File: apps/explorer/lib/explorer/chain/beacon/reader.ex (hunk 5 above)
Comment (vbaranov) [inline apps/explorer/lib/explorer/chain/beacon/reader.ex:46]: Should we check for consensus block in order to return only blobs from consensus blocks?

File: apps/indexer/lib/indexer/fetcher/beacon/client.ex
Comment (vbaranov) [inline apps/indexer/lib/indexer/fetcher/beacon/client.ex:1]: Please add spec & doc for public functions

File: apps/indexer/lib/indexer/fetcher/beacon/client.ex [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+            inspect(error, limit: :infinity, printable_limit: :infinity)
+          ]
+        end)
+
+        Logger.configure(truncate: old_truncate)
+        {:error, @request_error_msg}
+    end
+  end
+
+  def get_blob_sidecars(slots) when is_list(slots) do
```
Comment (vbaranov) [inline apps/indexer/lib/indexer/fetcher/beacon/client.ex:34]: Should we check that `slots` array is not empty?

File: config/runtime.exs [hunk 9]
```diff
@@ -679,6 +679,21 @@ config :indexer, Indexer.Fetcher.RootstockData,
...
+      ConfigHelper.parse_bool_env_var("INDEXER_DISABLE_BEACON_BLOB_SANITIZE_FETCHER")
+
+config :indexer, Indexer.Fetcher.Beacon.Blob,
+  slot_duration: ConfigHelper.parse_integer_env_var("INDEXER_BEACON_BLOB_SANITIZE_FETCHER_SLOT_DURATION", 12),
+  reference_slot: ConfigHelper.parse_integer_env_var("INDEXER_BEACON_BLOB_SANITIZE_FETCHER_REFERENCE_SLOT", 8_206_822),
+  reference_timestamp:
+    ConfigHelper.parse_integer_env_var("INDEXER_BEACON_BLOB_SANITIZE_FETCHER_REFERENCE_TIMESTAMP", 1_705_305_887),
+  start_block: ConfigHelper.parse_integer_env_var("INDEXER_BEACON_BLOB_SANITIZE_FETCHER_START_BLOCK", 8_206_822),
+  end_block: ConfigHelper.parse_integer_env_var("INDEXER_BEACON_BLOB_SANITIZE_FETCHER_END_BLOCK", 0)
+
```
Comment (vbaranov) [inline config/runtime.exs:696]: Please make a PR in order to update envs list in docs repo. And these envs should be added to `docker-compose/envs/common-blockscout.env` as well.

File: apps/explorer/test/support/factory.ex [hunk 10]
```diff
@@ -65,8 +65,8 @@ defmodule Explorer.Factory do
   end
 
   def auth_factory do
-    %Auth{
-      info: %Info{
+    %{
```
Comment (vbaranov) [inline apps/explorer/test/support/factory.ex:68]: Could you please elaborate what is the purpose of the change?

File: apps/block_scout_web/lib/block_scout_web/views/api/v2/transaction_view.ex (hunk 1 above)
Comment (k1rill-fedoseev) [reply apps/block_scout_web/lib/block_scout_web/views/api/v2/transaction_view.ex:453]: Did a small refactor in `suave_fields` to resolve the warning

File: apps/block_scout_web/lib/block_scout_web/views/api/v2/blob_view.ex (hunk 2 above)
Comment (k1rill-fedoseev) [reply apps/block_scout_web/lib/block_scout_web/views/api/v2/blob_view.ex]: Refactored using `Data.t()` type, helper is no longer needed

File: apps/explorer/test/support/factory.ex (hunk 10 above)
Comment (k1rill-fedoseev) [reply apps/explorer/test/support/factory.ex]: Reverted the changes, committed them by mistake

File: apps/explorer/lib/explorer/chain/beacon/reader.ex (hunk 5 above)
Comment (k1rill-fedoseev) [reply apps/explorer/lib/explorer/chain/beacon/reader.ex]: I think we can omit the consensus check here, since we are only interested in the block timestamps, which will be equal across all reorgs, so no more data than necessary will be fetched. In the worst case, we might rarely fetch some consensus blocks which don't contain any blobs at all, but I don't see it as a problem.

**Review by vbaranov — COMMENTED** [review]
I would suggest adding basic compatibility with the old UI too (at least not to crash) when searching by blob hashes:
```
Request: GET /search?q=0x01cdf8e590ba85883cbff001b73d7996d357e335f19fa896ed1a83aa477d952f
** (exit) an exception was raised:
    ** (FunctionClauseError) no function clause matching in BlockScoutWeb.ChainController.redirect_search_results/2
        (block_scout_web 6.1.0) lib/block_scout_web/controllers/chain_controller.ex:153: BlockScoutWeb.ChainController.redirect_search_results(%Plug.Conn{adapter: {Plug.Cowboy.Conn, :...}
```

File: apps/indexer/lib/indexer/fetcher/beacon/client.ex (hunk 8 above)
Comment (k1rill-fedoseev) [reply apps/indexer/lib/indexer/fetcher/beacon/client.ex:43]: Should work fine for empty lists as well. Added an implementation for empty list just in case.

File: apps/explorer/lib/explorer/chain/search.ex [hunk 11]
```diff
@@ -101,17 +102,28 @@ defmodule Explorer.Chain.Search do
...
-          user_operation_query = search_user_operation_query(string)
-
+        tx_block_query =
           basic_query
           |> union(^tx_query)
-          |> union(^user_operation_query)
           |> union(^block_query)
+
+        tx_block_op_query =
+          if UserOperation.user_operations_enabled?() do
```
[the remark this answers is not in the document]
Comment (k1rill-fedoseev) [reply apps/explorer/lib/explorer/chain/search.ex]: Renamed in b42f8e3

**Review by vbaranov — APPROVED** [review]

File: apps/explorer/priv/beacon/migrations/20240109102458_create_blobs_tables.exs [hunk 12]
```diff
@@ -0,0 +1,34 @@
+defmodule Explorer.Repo.Beacon.Migrations.CreateBlobsTables do
+  use Ecto.Migration
+
+  def change do
+    create table(:beacon_blobs_transactions, primary_key: false) do
+      add(:hash, references(:transactions, column: :hash, on_delete: :delete_all, type: :bytea),
```
Comment (varasev) [inline apps/explorer/priv/beacon/migrations/20240109102458_create_blobs_tables.exs:6]: Maybe rename this field to `transaction_hash`? @vbaranov do you agree?

**Comment by varasev** [comment]
In the PR description:

> * Add `/api/v2/:transaction_hash_param/blobs`, `/api/v2/blobs/:blob_hash_param`

I guess, you meant `/api/v2/transactions/:transaction_hash_param/blobs` here?

File: apps/block_scout_web/lib/block_scout_web/controllers/api/v2/transaction_controller.ex [hunk 13]
```diff
@@ -390,6 +403,23 @@ defmodule BlockScoutWeb.API.V2.TransactionController do
...
 
+  @doc """
+  Function to handle GET requests to `/api/v2/transactions/:transaction_hash_param/blobs` endpoint.
+  """
+  @spec blobs(Plug.Conn.t(), map()) :: Plug.Conn.t() | {atom(), any()}
+  def blobs(conn, %{"transaction_hash_param" => transaction_hash_string} = params) do
+    with {:ok, _transaction, transaction_hash} <- validate_transaction(transaction_hash_string, params) do
+      full_options = @api_true
+
+      blobs = Chain.transaction_to_blobs(transaction_hash, full_options)
```
Comment (varasev) [inline apps/block_scout_web/lib/block_scout_web/controllers/api/v2/transaction_controller.ex:414]: I'd suggest to move the blob-related functions from the `Explorer.Chain` to the `Explorer.Chain.Blobs` because we aim to reduce the size of `apps/explorer/lib/explorer/chain.ex` file

Comment (k1rill-fedoseev) [reply apps/block_scout_web/lib/block_scout_web/controllers/api/v2/transaction_controller.ex]: moved it to the file with the rest of the blob-related selects

File: apps/explorer/priv/beacon/migrations/20240109102458_create_blobs_tables.exs (hunk 12 above)
Comment (k1rill-fedoseev) [reply apps/explorer/priv/beacon/migrations/20240109102458_create_blobs_tables.exs:6]: My intention behind naming it `hash` was so it's gonna be easier to understand that it's a primary key. `transaction_hash` usually looks like a non-primary foreign key.
